File: app/src/main/java/com/soya/launcher/ui/dialog/KeyboardDialog.java
package com.soya.launcher.ui.dialog;

import android.content.DialogInterface;
import android.os.Bundle;
import android.util.Log;
import android.view.Gravity;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.soya.launcher.R;
import com.soya.launcher.adapter.KeyboardAdapter;
import com.soya.launcher.bean.KeyItem;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class KeyboardDialog extends SingleDialogFragment implements KeyboardAdapter.Callback {
    public static final String TAG = "KeyboardDialog";
    public static KeyboardDialog newInstance() {

        Bundle args = new Bundle();

        KeyboardDialog fragment = new KeyboardDialog();
        fragment.setArguments(args);
        return fragment;
    }

    private RecyclerView mRecyclerView;
    private KeyboardAdapter mAdapter;
    private Callback callback;

    private final List<KeyItem> ens = new ArrayList<>();
    private final List<KeyItem> nums = new ArrayList<>();

    private TextView mTargetView;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        fillEn();
        fillNums();
    }

    @Override
    protected int getLayout() {
        return R.layout.dialog_keyboard;
    }

    @Override
    protected void init(LayoutInflater inflater, View view) {
        super.init(inflater, view);
        mRecyclerView = view.findViewById(R.id.recycler);

        mAdapter = new KeyboardAdapter(getActivity(), inflater, new ArrayList<>());
    }

    @Override
    protected void initBefore(LayoutInflater inflater, View view) {
        super.initBefore(inflater, view);
        mAdapter.setCallback(this);
        getDialog().setOnKeyListener(new DialogInterface.OnKeyListener() {
            @Override
            public boolean onKey(DialogInterface dialog, int keyCode, KeyEvent event) {
                if (event.getAction() == KeyEvent.ACTION_UP){
                    if (event.getUnicodeChar() == 0){
                        switch (keyCode){
                            case KeyEvent.KEYCODE_DEL:
                                del();
                                break;
                        }
                    }else {
                        mTargetView.append(String.valueOf(event.getDisplayLabel()));
                    }
                }
                return false;
            }
        });
    }

    @Override
    protected void initBind(LayoutInflater inflater, View view) {
        super.initBind(inflater, view);
        mAdapter.replace(ens, KeyboardAdapter.TYPE_ENG);
        mRecyclerView.setLayoutManager(new GridLayoutManager(getActivity(), 10));
        mRecyclerView.setAdapter(mAdapter);
    }

    @Override
    protected int getGravity() {
        return Gravity.BOTTOM | Gravity.CENTER_HORIZONTAL;
    }

    @Override
    protected int[] getWidthAndHeight() {
        return new int[]{RecyclerView.LayoutParams.MATCH_PARENT, RecyclerView.LayoutParams.MATCH_PARENT};
    }

    @Override
    protected float getDimAmount() {
        return 0;
    }

    public void setTargetView(TextView targetView) {
        this.mTargetView = targetView;
    }

    private void fillEn(){
        String[] array = "q w e r t y u i o p".split(" ");
        Log.e(TAG, "fillEn: "+ Arrays.toString(array));
        for (String item : array) ens.add(new KeyItem(KeyItem.TYPE_ENG, item, -1, 1, false));

        array = "a s d f g h j k l ,".split(" ");
        for (String item : array) ens.add(new KeyItem(KeyItem.TYPE_ENG, item, -1, 1, false));

        ens.add(new KeyItem(KeyItem.TYPE_UPCAST, "", R.drawable.baseline_lower_cast_100, 1, true));
        array = "z x c v b n m .".split(" ");
        for (String item : array) ens.add(new KeyItem(KeyItem.TYPE_ENG, item, -1, 1, false));
        ens.add(new KeyItem(KeyItem.TYPE_DEL, "", R.drawable.baseline_backspace_100, 1, true));

        ens.add(new KeyItem(KeyItem.TYPE_SWITCH, "?123", -1, 2, false));
        ens.add(new KeyItem(KeyItem.TYPE_SPACE, " ", -1, 6, false));
        ens.add(new KeyItem(KeyItem.TYPE_SEARCH, "", R.drawable.baseline_search_100_2, 2, true));
    }

    private void fillNums(){

        String[] array = "1 2 3 4 5 6 7 8 9 0".split(" ");
        for (String item : array) nums.add(new KeyItem(KeyItem.TYPE_NUM, item, -1, 1, false));

        array = "! @ # $ % ^ & * ( )".split(" ");
        for (String item : array) nums.add(new KeyItem(KeyItem.TYPE_ENG, item, -1, 1, false));

        array = "~ , . ? / \" - + =".split(" ");
        for (String item : array) nums.add(new KeyItem(KeyItem.TYPE_ENG, item, -1, 1, false));
        nums.add(new KeyItem(KeyItem.TYPE_DEL, "", R.drawable.baseline_backspace_100, 1, true));

        nums.add(new KeyItem(KeyItem.TYPE_SWITCH, "ABC", -1, 2, false));
        nums.add(new KeyItem(KeyItem.TYPE_SPACE, " ", -1, 6, false));
        nums.add(new KeyItem(KeyItem.TYPE_SEARCH, "", R.drawable.baseline_search_100_2, 2, true));
    }

    @Override
    public void onClick(KeyItem bean, String text) {
        switch (bean.getType()){
            case KeyItem.TYPE_SWITCH:
                if (mAdapter.getType() == KeyboardAdapter.TYPE_ENG){
                    mAdapter.replace(nums, KeyboardAdapter.TYPE_NUM);
                }else if (mAdapter.getType() == KeyboardAdapter.TYPE_NUM){
                    mAdapter.replace(ens, KeyboardAdapter.TYPE_ENG);
                }
                break;
            case KeyItem.TYPE_UPCAST:
                mAdapter.setUPCase(!mAdapter.isUPCase());
                break;
            case KeyItem.TYPE_DEL:
                del();
                break;
            case KeyItem.TYPE_SEARCH:
                mTargetView.onEditorAction(EditorInfo.IME_ACTION_DONE);
                break;
            default:
                mTargetView.append(text);
        }
    }

    private void del(){
        int len = mTargetView.getText().length();
        if (len != 0) mTargetView.setText(mTargetView.getText().subSequence(0, len - 1));
    }

    @Override
    public void onDismiss(@NonNull DialogInterface dialog) {
        super.onDismiss(dialog);
        if (callback != null) callback.onClose();
    }

    public void setCallback(Callback callback) {
        this.callback = callback;
    }

    public interface Callback{
        void onClose();
    }
}

File: app/src/main/java/com/soya/launcher/ui/dialog/KeyboardDialog.kt
package com.soya.launcher.ui.dialog

import android.content.DialogInterface
import android.os.Bundle
import android.util.Log
import android.view.Gravity
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.TextView
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.shudong.lib_base.ext.d
import com.shudong.lib_base.ext.e
import com.shudong.lib_base.ext.no
import com.shudong.lib_base.ext.otherwise
import com.shudong.lib_base.ext.yes
import com.soya.launcher.R
import com.soya.launcher.adapter.KeyboardAdapter
import com.soya.launcher.bean.KeyItem
import com.soya.launcher.ext.isAndroidAtMost5_1
import com.soya.launcher.utils.toTrim

class KeyboardDialog : SingleDialogFragment(), KeyboardAdapter.Callback {
    private var mRecyclerView: RecyclerView? = null
    private var mAdapter: KeyboardAdapter? = null
    private var callback: Callback? = null
    private val ens: MutableList<KeyItem> = ArrayList()
    private val nums: MutableList<KeyItem> = ArrayList()
    private var mTargetView: TextView? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        fillEn()
        fillNums()
    }

    override fun getLayout(): Int {
        return R.layout.dialog_keyboard
    }

    override fun init(inflater: LayoutInflater, view: View) {
        super.init(inflater, view)
        mRecyclerView = view.findViewById(R.id.recycler)
        mAdapter = KeyboardAdapter(requireContext(), inflater, ArrayList())
    }

    var isPressSpace = false
    override fun initBefore(inflater: LayoutInflater, view: View) {
        super.initBefore(inflater, view)
        mAdapter!!.setCallback(this)
        dialog!!.setOnKeyListener { dialog, keyCode, event ->
            if (event.action == KeyEvent.ACTION_UP) {
                if (event.unicodeChar == 0) {
                    when (keyCode) {
                        KeyEvent.KEYCODE_DEL -> del()
                    }
                } else {
                    when (keyCode) {
                        KeyEvent.KEYCODE_SPACE -> isPressSpace = true
                        else-> isPressSpace = false
                    }
                    mTargetView!!.append(event.displayLabel.toString())
                }
            }
            false
        }
    }

    override fun initBind(inflater: LayoutInflater, view: View) {
        super.initBind(inflater, view)
        mAdapter!!.replace(ens, KeyboardAdapter.TYPE_ENG)
        mRecyclerView!!.setLayoutManager(GridLayoutManager(activity, 10))
        mRecyclerView!!.setAdapter(mAdapter)
    }

    override fun getGravity(): Int {
        return Gravity.BOTTOM or Gravity.CENTER_HORIZONTAL
    }

    override fun getWidthAndHeight(): IntArray {
        return intArrayOf(
            RecyclerView.LayoutParams.MATCH_PARENT,
            RecyclerView.LayoutParams.MATCH_PARENT
        )
    }

    override fun getDimAmount(): Float {
        return 0f
    }

    fun setTargetView(targetView: TextView?) {
        mTargetView = targetView
    }

    private fun fillEn() {
        var array =
            "q w e r t y u i o p".split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

        for (item in array) ens.add(KeyItem(KeyItem.TYPE_ENG, item, -1, 1, false))
        array =
            "a s d f g h j k l ,".split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        for (item in array) ens.add(KeyItem(KeyItem.TYPE_ENG, item, -1, 1, false))
        ens.add(KeyItem(KeyItem.TYPE_UPCAST, "", R.drawable.baseline_lower_cast_100, 1, true))
        array = "z x c v b n m .".split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        for (item in array) ens.add(KeyItem(KeyItem.TYPE_ENG, item, -1, 1, false))
        ens.add(KeyItem(KeyItem.TYPE_DEL, "", R.drawable.baseline_backspace_100, 1, true))
        ens.add(KeyItem(KeyItem.TYPE_SWITCH, "?123", -1, 2, false))
        ens.add(KeyItem(KeyItem.TYPE_SPACE, " ", -1, 6, false))
        ens.add(KeyItem(KeyItem.TYPE_SEARCH, "DONE", -1, 2, true))
    }

    private fun fillNums() {
        var array =
            "1 2 3 4 5 6 7 8 9 0".split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        for (item in array) nums.add(KeyItem(KeyItem.TYPE_NUM, item, -1, 1, false))
        array =
            "! @ # $ % ^ & * ( )".split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        for (item in array) nums.add(KeyItem(KeyItem.TYPE_ENG, item, -1, 1, false))
        array =
            "~ , . ? / \" - + =".split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        for (item in array) nums.add(KeyItem(KeyItem.TYPE_ENG, item, -1, 1, false))
        nums.add(KeyItem(KeyItem.TYPE_DEL, "", R.drawable.baseline_backspace_100, 1, true))
        nums.add(KeyItem(KeyItem.TYPE_SWITCH, "ABC", -1, 2, false))
        nums.add(KeyItem(KeyItem.TYPE_SPACE, " ", -1, 6, false))
        nums.add(KeyItem(KeyItem.TYPE_SEARCH, "DONE", -1, 2, true))
    }

    override fun onClick(bean: KeyItem, text: String) {
        when (bean.type) {
            KeyItem.TYPE_SWITCH -> if (mAdapter!!.type == KeyboardAdapter.TYPE_ENG) {
                mAdapter!!.replace(nums, KeyboardAdapter.TYPE_NUM)
            } else if (mAdapter!!.type == KeyboardAdapter.TYPE_NUM) {
                mAdapter!!.replace(ens, KeyboardAdapter.TYPE_ENG)
            }

            KeyItem.TYPE_UPCAST -> mAdapter!!.setUPCase(!mAdapter!!.isUPCaseKey)
            KeyItem.TYPE_DEL -> {
                "按下了删除=="
                del()
            }
            KeyItem.TYPE_SEARCH -> {
                "按下了搜索=="
                mTargetView!!.onEditorAction(EditorInfo.IME_ACTION_DONE)
                dismiss()
            }

            KeyItem.TYPE_SPACE->{
                "按下了空格=="
                mTargetView!!.append(" ")
            }
            else -> {
                isPressSpace.no {
                    isAndroidAtMost5_1().yes {
                        //mTargetView!!.append(text.toTrim())
                        mTargetView?.text = mTargetView?.text.toString().toTrim() + text.toTrim()
                    }.otherwise {
                        mTargetView!!.append(text)
                        //mTargetView?.text = mTargetView?.text?.toString()
                    }
                }



            }
        }
    }

    private fun del() {

        isAndroidAtMost5_1().yes { mTargetView?.text = mTargetView?.text?.toString()?.toTrim() }


        val len = mTargetView?.text?.length
        if (len != 0) {
            var finalStr = ""
            isAndroidAtMost5_1().yes {
                val str = mTargetView?.text?.toString()?.toTrim()
                val len1 = (mTargetView?.text?.toString()?.toTrim()?.length?:0).minus(1)
                 finalStr = str?.subSequence(0, len1).toString()
            }.otherwise {
                val str = mTargetView?.text?.toString()
                val len1 = (mTargetView?.text?.toString()?.length?:0).minus(1)
                finalStr = str?.subSequence(0, len1).toString()
            }


           // "设置之后,长文字是====${str}===${finalStr}===$len1"
            mTargetView?.text = finalStr

        }
    }

    override fun onDismiss(dialog: DialogInterface) {
        super.onDismiss(dialog)
        if (callback != null) callback!!.onClose()
    }

    fun setCallback(callback: Callback?) {
        this.callback = callback
    }

    interface Callback {
        fun onClose()
    }

    companion object {
        const val TAG = "KeyboardDialog"
        @JvmStatic
        fun newInstance(): KeyboardDialog {
            val args = Bundle()
            val fragment = KeyboardDialog()
            fragment.setArguments(args)
            return fragment
        }
    }
}
